### PR TITLE
Fix scalar write to event file

### PR DESCRIPTION
**Pull request by vandanavk**
### Description of changes:
save_scalar wasn't writing events because of a check in close_Writers(). This PR contains a fix for that along with the corresponding fix to the unit test.
Apart from these changes, this PR contains 
1. doc update
2. additional tests for TF SessionHook and XGBoost.
3. enabling save_scalar for TF
4. init_step =-1 for Keras, adding a check in _on_any_batch_begin() (similar to the checks in pytorch and mxnet) to start writing after the 1st batch.
5. Fix to TF close_writers() to flush tb_writers -> Tested by checking the tensorboard output of test_tf_save_scalar(). screenshot of tb output pasted below.

Output:
PyTorch: Loss and custom scalar
Screen Shot 2020-01-14 at 5 43 51 PM

MXNet index file:
`{"meta": {"mode": "TRAIN", "mode_step": 0, "event_file_name": "events/000000000000/000000000000_worker_0.tfevents"}, "tensor_payload": [{"tensorname": "softmaxcrossentropyloss0_output_0", "start_idx": 0, "length": 275}, {"tensorname": "conv0_weight", "start_idx": 275, "length": 764}, {"tensorname": "gradient/conv0_weight", "start_idx": 1039, "length": 809}, {"tensorname": "conv0_bias", "start_idx": 1848, "length": 120}, {"tensorname": "gradient/conv0_bias", "start_idx": 1968, "length": 138}, {"tensorname": "conv1_weight", "start_idx": 2106, "length": 3621}, {"tensorname": "gradient/conv1_weight", "start_idx": 5727, "length": 3666}, {"tensorname": "conv1_bias", "start_idx": 9393, "length": 163}, {"tensorname": "gradient/conv1_bias", "start_idx": 9556, "length": 181}, {"tensorname": "dense0_weight", "start_idx": 9737, "length": 192136}, {"tensorname": "gradient/dense0_weight", "start_idx": 201873, "length": 192163}, {"tensorname": "dense0_bias", "start_idx": 394036, "length": 584}, {"tensorname": "gradient/dense0_bias", "start_idx": 394620, "length": 602}, {"tensorname": "dense1_weight", "start_idx": 395222, "length": 40454}, {"tensorname": "gradient/dense1_weight", "start_idx": 435676, "length": 40481}, {"tensorname": "dense1_bias", "start_idx": 476157, "length": 439}, {"tensorname": "gradient/dense1_bias", "start_idx": 476596, "length": 457}, {"tensorname": "dense2_weight", "start_idx": 477053, "length": 3489}, {"tensorname": "gradient/dense2_weight", "start_idx": 480542, "length": 3516}, {"tensorname": "dense2_bias", "start_idx": 484058, "length": 139}, {"tensorname": "gradient/dense2_bias", "start_idx": 484197, "length": 159}, {"tensorname": "scalar/mx_num_steps", "start_idx": 484356, "length": 122}, {"tensorname": "scalar/mx_before_train", "start_idx": 484478, "length": 128}]}`

Screen Shot 2020-01-14 at 5 11 03 PM

Note: MXNet loss is not written to tensorboard as it is a vector. 

XGBoost:
Screen Shot 2020-01-15 at 2 30 01 PM

TF Keras index file:
`{"meta": {"mode": "TRAIN", "mode_step": 0, "event_file_name": "events/000000000000/000000000000_worker_0.tfevents"}, "tensor_payload": [{"tensorname": "dense/weights/dense/kernel:0", "start_idx": 0, "length": 401590}, {"tensorname": "dense/weights/dense/bias:0", "start_idx": 401590, "length": 647}, {"tensorname": "dense_1/weights/dense_1/kernel:0", "start_idx": 402237, "length": 5308}, {"tensorname": "dense_1/weights/dense_1/bias:0", "start_idx": 407545, "length": 179}, {"tensorname": "dense_1/outputs/dense_1/Softmax:0", "start_idx": 407724, "length": 2400202}, {"tensorname": "training/gradients/dense/BiasAdd_grad/tuple/control_dependency_1:0", "start_idx": 2807926, "length": 727}, {"tensorname": "training/gradients/dense/MatMul_grad/tuple/control_dependency_1:0", "start_idx": 2808653, "length": 401702}, {"tensorname": "training/gradients/dense_1/BiasAdd_grad/tuple/control_dependency_1:0", "start_idx": 3210355, "length": 255}, {"tensorname": "training/gradients/dense_1/MatMul_grad/tuple/control_dependency_1:0", "start_idx": 3210610, "length": 5414}, {"tensorname": "batch", "start_idx": 3216024, "length": 94}, {"tensorname": "size", "start_idx": 3216118, "length": 92}, {"tensorname": "acc", "start_idx": 3216210, "length": 86}, {"tensorname": "loss", "start_idx": 3216296, "length": 88}, {"tensorname": "scalar/tf_keras_num_steps", "start_idx": 3216384, "length": 134}, {"tensorname": "scalar/tf_keras_before_train", "start_idx": 3216518, "length": 140}]}`

Screen Shot 2020-01-15 at 5 31 24 PM

TF Session:
Screen Shot 2020-01-15 at 6 26 50 PM

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/core/test_hook_save_scalar.py [hunk 1]
```diff
@@ -277,8 +343,7 @@ def check_trials(out_dir, save_steps, saved_scalars=None):
             if eval_steps:  # need this check for bias and gradients
                 assert len(set(save_steps["EVAL"]) & set(eval_steps)) == len(save_steps["EVAL"])
     scalar_list = trial.tensor_names(regex="^scalar")
-    if scalar_list:
-        assert len(set(saved_scalars) & set(scalar_list)) == len(saved_scalars)
+    assert len(set(saved_scalars) & set(scalar_list)) == len(saved_scalars)
```
Comment (Vikas-kum) [inline tests/core/test_hook_save_scalar.py]: if saved_scalars then

File: tests/core/test_hook_save_scalar.py [hunk 2]
```diff
@@ -343,7 +343,8 @@ def check_trials(out_dir, save_steps, saved_scalars=None):
             if eval_steps:  # need this check for bias and gradients
                 assert len(set(save_steps["EVAL"]) & set(eval_steps)) == len(save_steps["EVAL"])
     scalar_list = trial.tensor_names(regex="^scalar")
-    assert len(set(saved_scalars) & set(scalar_list)) == len(saved_scalars)
+    if saved_scalars:
+        assert len(set(saved_scalars) & set(scalar_list)) == len(saved_scalars)
```
Comment (Vikas-kum) [inline tests/core/test_hook_save_scalar.py]: why this special intersection ? 
Shouldn't len(Saved_scalars) == len(scalar_list) 

What is the reason where one list can be more than other list?

**Comment by Vikas-kum** [comment]
can you update the description with testing done and output for TF, Keras, Xgboost also

**Comment by jarednielsen** [comment]
Thanks for catching the tensorboard write!

File: smdebug/tensorflow/base_hook.py [hunk 3]
```diff
@@ -339,6 +339,16 @@ def _close_writers(self) -> None:
...
 
+        to_delete_writers = []
+        # Delete all the tb writers
+        for mode, writer in self.tb_writers.items():
+            if writer is not None:
+                writer.flush()
+                writer.close()
+                to_delete_writers.append(mode)
+        for mode in to_delete_writers:
+            del self.tb_writers[mode]
```
Comment (Vikas-kum) [inline smdebug/tensorflow/base_hook.py]: Why do we need to explicitly delete these?

Comment (vandanavk) [reply smdebug/tensorflow/base_hook.py]: all writers are closed and deleted in close_writers() marking the end of a file. initialize_writers() or maybe_get_tb_writer() creates a new writer for the next file - in this function, there is a check to see if a writer exists for a particular mode. if there is no writer for a mode, it creates a new one.
probably done this way (rather than just setting the writer to None) to avoid a scenario where a write is made using a None writer.
